### PR TITLE
[SILGen] Emit default arg generator for stored property kind

### PR DESCRIPTION
**Pull request by Azoy**
In order to properly serialize stored property default arguments, we need to emit a default arg generator.

I chose to emit the parent initializer expression instead of the apply to the variable initializer because SIL would eventually just inline the function definition anyway (or at least in most cases from my findings) and this was a lot cleaner and cheaper implementation wise. I've updated the SILGen tests and added a serialization test.

cc: @slavapestov @jrose-apple

### CODE REVIEW

**Comment by jrose-apple** [comment]
The approach looks good to me but I think Slava should check the details. Thank you, Alejandro!

**Comment by Azoy** [comment]
I'm questioning whether whether we need a special default arg kind anymore with this implementation. Also, I feel like I explored this route in the initial patch, but just couldn't make it work then.

**Review by slavapestov — COMMENTED** [review]
I’m a bit nervous about this because now we will emit some expressions twice, once for the property initializer and once for the default argument. In the past we’ve had bugs with multiple emission of expressions; that’s why we introduced property initializers in the first place, so that each constructor doesn’t re-emit these expressions. I would suggest therefore changing the default argument generators to call the property initializers instead of re-emitting them, even though it’s slightly more complicated.

**Comment by Azoy** [comment]
Oh I didn't know this was the case for expression. 👍

**Comment by Azoy** [comment]
@slavapestov I believe I did this correctly, but would love feedback. I've added some more tests to exercise indirect results.

File: lib/SILGen/SILGenFunction.cpp [hunk 1]
```diff
@@ -638,6 +638,54 @@ void SILGenFunction::emitGeneratorFunction(SILDeclRef function, Expr *value) {
...
 
+void SILGenFunction::emitGeneratorFunction(SILDeclRef function, VarDecl *var) {
+  MagicFunctionName = SILGenModule::getMagicFunctionName(function);
+
+  RegularLocation loc(var);
+  loc.markAutoGenerated();
+
+  auto decl = function.getAbstractFunctionDecl();
+  auto *dc = decl->getInnermostDeclContext();
+  auto interfaceType = var->getInterfaceType();
```
Comment (slavapestov) [inline lib/SILGen/SILGenFunction.cpp]: Can you test weak or unowned properties? I suspect you might need to change this to getValueInterfaceType() to strip off the reference storage qualifier.

**Review by slavapestov — APPROVED** [review]
Looks good!

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by Azoy** [comment]
@slavapestov okay, I believe I have fixed the TBD test, as well as pull in your test changes as well.

**Review by slavapestov — APPROVED** [review]

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by jrose-apple** [comment]
PromiseKit failure looks real. :-( https://ci.swift.org/job/swift-PR-source-compat-suite-debug/976/artifact/swift-source-compat-suite/FAIL_PromiseKit-PromiseKit.xcodeproj_4.0.3_BuildXcodeProjectTarget_PromiseKit_generic-platform-macOS.log

**Comment by Azoy** [comment]
I don’t believe the Linux failure is related, but I’m not sure why debug source compat failed, but release didn’t?

**Comment by jrose-apple** [comment]
Release is usually a whole-module build; maybe you have a default arg generator being emitted from one file that references an initial value in another file, or something like that?

**Comment by Azoy** [comment]
Thanks Jordan! I think I see the issue, will work on this when I get home.

**Comment by Azoy** [comment]
@slavapestov I hate to ping you again, but I believe this is the one!

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test macOS

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
Thanks @Azoy!

**Pull request merged**